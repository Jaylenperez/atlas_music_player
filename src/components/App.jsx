import Footer from "./Footer";
import MusicPlayer from "./MusicPlayer";

function App() {
  return (
    <div className="flex h-full min-h-screen flex-col justify-between bg-amber-100 p-8 dark:bg-purple-950">
      <MusicPlayer />
      <Footer />
    </div>
  );
}

export default App;
