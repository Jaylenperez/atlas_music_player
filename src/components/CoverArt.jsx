import placeholder from "../assets/placeholder.svg";

export default function CoverArt() {
  return (
    <figure className="aspect-square w-full overflow-hidden rounded-md bg-neutral-200 shadow ring-1 ring-black/10 dark:bg-neutral-800 dark:ring-white/10">
      <img
        src={placeholder}
        alt="Album cover art"
        className="h-full w-full object-cover"
        loading="lazy"
      />
    </figure>
  );
}
