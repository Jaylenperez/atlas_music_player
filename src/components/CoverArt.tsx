import React from "react";

export interface CoverArtProps {
  src: string;
  alt?: string;
}

export const CoverArt: React.FC<CoverArtProps> = ({ src, alt = "Cover Art" }) => {
  return (
    <div className="flex justify-center md:justify-start">
      <img
        src={src}
        alt={alt}
        className="aspect-square w-full rounded-md object-cover sm:w-72 md:w-80 lg:w-96 xl:w-[50rem]"
      />
    </div>
  );
};
