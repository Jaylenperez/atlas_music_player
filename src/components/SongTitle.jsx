export default function SongTitle() {
  return (
    <div>
      <h1 className="text-plum font-sans text-3xl font-bold">Song Title</h1>
      <p className="text-accent font-sans text-lg font-bold text-gray-500">
        Song Artist
      </p>
    </div>
  );
}
