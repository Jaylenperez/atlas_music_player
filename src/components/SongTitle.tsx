import React from "react";

export interface SongTitleProps {
  lightMode: boolean;
  title: string;
  artist: string;
}

export const SongTitle: React.FC<SongTitleProps> = ({
  lightMode,
  title,
  artist,
}) => (
  <div className="w-full">
    <h2
      className={`text-left text-2xl font-bold ${
        lightMode ? "text-indigo-950" : "text-amber-200"
      }`}
    >
      {title}
    </h2>
    <p
      className={`text-left text-lg ${
        lightMode ? "text-gray-500" : "text-gray-300"
      }`}
    >
      {artist}
    </p>
  </div>
);
