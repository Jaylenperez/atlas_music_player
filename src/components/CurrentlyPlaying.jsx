import CoverArt from "./CoverArt";
import SongTitle from "./SongTitle";
import PlayControls from "./PlayControls";
import VolumeControls from "./VolumeControls";

export default function CurrentlyPlaying() {
  return (
    <div className="flex flex-col items-center gap-4">
      <CoverArt />

      <div className="w-full max-w-[400px] text-left">
        <SongTitle />
      </div>

      <div className="text-plum">
        <PlayControls />
      </div>

      <VolumeControls />
    </div>
  );
}
