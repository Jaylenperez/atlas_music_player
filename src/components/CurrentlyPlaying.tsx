import React, { useEffect, useRef, useState } from "react";
import { CoverArt } from "./CoverArt";
import { SongTitle } from "./SongTitle";
import PlayControls from "./PlayControls";
import { VolumeControls } from "./VolumeControls";
import { PlaylistSong } from "./App";

interface Song extends PlaylistSong {
  cover: string;
  song: string;
}

interface Props {
  lightMode: boolean;
  track: PlaylistSong;
  playlist: PlaylistSong[];
  onSelect: (t: PlaylistSong) => void;
}

const CurrentlyPlaying: React.FC<Props> = ({
  lightMode,
  track,
  playlist,
  onSelect,
}) => {
  const [fullSong, setFullSong] = useState<Song | null>(null);
  const [shuffleEnabled, setShuffleEnabled] = useState(false);
  const [volume, setVolume] = useState(50);
  const [isPlaying, setIsPlaying] = useState(false);
  const [playbackSpeed, setPlaybackSpeed] = useState<0.5 | 1 | 2>(1);

  const audioRef = useRef<HTMLAudioElement | null>(null);

  // Load song details
  useEffect(() => {
    fetch(`/api/v1/songs/${track.id}`)
      .then((r) => {
        if (!r.ok) throw new Error("Song fetch failed");
        return r.json();
      })
      .then((data: Song) => setFullSong(data))
      .catch(() => setFullSong(null));
  }, [track.id]);

  // When fullSong changes, swap audio
  useEffect(() => {
    if (!fullSong) return;
    audioRef.current?.pause();
    const audio = new Audio(fullSong.song);
    audio.volume = volume / 100;
    audio.playbackRate = playbackSpeed;   // set initial speed
    audioRef.current = audio;
    audio.play().then(() => setIsPlaying(true)).catch(() => {});
    return () => {
      audio.pause();
    };
  }, [fullSong]);

  // Update volume live
  useEffect(() => {
    if (audioRef.current) {
      audioRef.current.volume = volume / 100;
    }
  }, [volume]);

  // Update playbackRate when speed changes
  useEffect(() => {
    if (audioRef.current) {
      audioRef.current.playbackRate = playbackSpeed;
    }
  }, [playbackSpeed]);

  // Playback handlers
  const togglePlay = () => {
    if (!audioRef.current) return;
    if (isPlaying) {
      audioRef.current.pause();
      setIsPlaying(false);
    } else {
      audioRef.current
        .play()
        .then(() => setIsPlaying(true))
        .catch(() => {});
    }
  };

  // Back / Forward / Shuffle logic
  const idx = playlist.findIndex((t) => t.id === track.id);
  const isFirst = idx <= 0;
  const isLast = idx === playlist.length - 1;

  const handleBack = () => {
    if (!isFirst) onSelect(playlist[idx - 1]);
  };
  const handleForward = () => {
    if (shuffleEnabled) {
      const rand = Math.floor(Math.random() * playlist.length);
      onSelect(playlist[rand]);
    } else if (!isLast) {
      onSelect(playlist[idx + 1]);
    }
  };
  const toggleShuffle = () => setShuffleEnabled((s) => !s);

  const coverSrc = fullSong?.cover ?? "/placeholder.svg";

  return (
    <div className="flex h-full w-full flex-col justify-between px-4 py-4">
      <div className="flex-shrink-0">
        <CoverArt src={coverSrc} />
      </div>

      {fullSong && (
        <div className="space-y-3">
          <SongTitle
            lightMode={lightMode}
            title={fullSong.title}
            artist={fullSong.artist}
          />
          <PlayControls
            lightMode={lightMode}
            isPlaying={isPlaying}
            onPlayToggle={togglePlay}
            onBack={handleBack}
            onForward={handleForward}
            disableBack={isFirst}
            disableForward={!shuffleEnabled && isLast}
            shuffleEnabled={shuffleEnabled}
            onShuffleToggle={toggleShuffle}
            playbackSpeed={playbackSpeed}
            onSpeedChange={setPlaybackSpeed}
          />
        </div>
      )}

      <div className="flex-shrink-0">
        <VolumeControls
          lightMode={lightMode}
          volume={volume}
          onVolumeChange={setVolume}
        />
      </div>
    </div>
  );
};

export default CurrentlyPlaying;
