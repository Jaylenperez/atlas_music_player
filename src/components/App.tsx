import React, { useState, useEffect } from "react";
import CurrentlyPlaying from "./CurrentlyPlaying";
import Playlist from "./Playlist";

export interface PlaylistSong {
  id: string;
  title: string;
  artist: string;
  genre: string;
  duration: number;
}

const App: React.FC = () => {
  const [lightMode, setLightMode] = useState<boolean>(false);
  const [playlist, setPlaylist] = useState<PlaylistSong[]>([]);
  const [selectedTrack, setSelectedTrack] = useState<PlaylistSong | null>(null);

  useEffect(() => {
    fetch("/api/v1/playlist")
      .then((res) => res.json())
      .then((data: PlaylistSong[]) => {
        setPlaylist(data);
        if (data.length > 0) setSelectedTrack(data[0]);
      })
      .catch((err) => console.error("Failed to load playlist:", err));
  }, []);

  return (
    <div
      className={`${
        lightMode ? "bg-amber-100 text-gray-900" : "bg-indigo-950 text-white"
      } flex h-screen flex-col overflow-x-hidden overflow-y-auto md:overflow-y-hidden md:flex-row`}
    >
      <button
        onClick={() => setLightMode((m) => !m)}
        className={`absolute top-4 right-4 rounded px-3 py-1 text-sm font-medium transition ${
          lightMode
            ? "bg-slate-400 text-gray-800"
            : "bg-gray-800 text-gray-200"
        }`}
      >
        {lightMode ? "Dark Mode" : "Light Mode"}
      </button>

      <div className="flex w-full flex-col justify-between p-4 md:w-1/2 md:p-8">
        {selectedTrack && (
          <CurrentlyPlaying
            lightMode={lightMode}
            track={selectedTrack}
            playlist={playlist}
            onSelect={setSelectedTrack}
          />
        )}
      </div>

      <div className="w-full p-4 md:w-1/2 md:border-t-0 md:p-8 md:overflow-y-auto">
        <Playlist
          lightMode={lightMode}
          tracks={playlist}
          selected={selectedTrack}
          onSelect={setSelectedTrack}
        />
      </div>
    </div>
  );
};

export default App;
