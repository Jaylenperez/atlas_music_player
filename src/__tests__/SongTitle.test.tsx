import { render } from "@testing-library/react";
import { expect, test } from "vitest";
import { SongTitle } from "../components/SongTitle";


test("SongTitle renders correctly", () => {
    const { container } = render(
        <SongTitle
            songName="Radio"
            artistName="Lana Del Rey"
        />
    );
    expect(container).toMatchSnapshot();
});