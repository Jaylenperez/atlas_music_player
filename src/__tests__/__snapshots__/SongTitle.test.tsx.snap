// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SongTitle renders correctly 1`] = `
<div>
  <div
    class="w-full"
  >
    <h2
      class="text-left text-2xl font-bold text-amber-200"
    />
    <p
      class="text-left text-lg text-gray-300"
    />
  </div>
</div>
`;

exports[`SongTitle renders correctly with lightMode = false 1`] = `
<div>
  <div
    class="w-full"
  >
    <h2
      class="text-left text-2xl font-bold text-amber-200"
    >
      Another Song
    </h2>
    <p
      class="text-left text-lg text-gray-300"
    >
      Another Artist
    </p>
  </div>
</div>
`;

exports[`SongTitle renders correctly with lightMode = true 1`] = `
<div>
  <div
    class="w-full"
  >
    <h2
      class="text-left text-2xl font-bold text-indigo-950"
    >
      Song Title
    </h2>
    <p
      class="text-left text-lg text-gray-500"
    >
      Artist Name
    </p>
  </div>
</div>
`;
